core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -390,7 +394,16 @@ protected void validateDataFilesExist(TableMetadata base, Long startingSnapshotI
   @Override
   public Object updateEvent() {
     long snapshotId = snapshotId();
-    long sequenceNumber = ops.refresh().snapshot(snapshotId).sequenceNumber();
+    Snapshot justSaved = ops.refresh().snapshot(snapshotId);
+    long sequenceNumber = 0;
+    if( justSaved == null) {
+      // The snapshot just saved may not be present if the latest metadata couldn't be loaded due to eventual
+      // consistency problems in refresh.
+      LOG.warn("Failed to load committed snapshot, leave sequence number to 0");
+    } else {
+      sequenceNumber = justSaved.sequenceNumber();
```

Comment:
I think the fix looks good but I'm not sure if we should default `sequenceNumber` to 0, as this may result in incorrect data for users that rely on this sequence number in the event. Since without this change we probably will not send out the event at all, if we don't want to wait until the data become eventual I wonder if returning -1 would be better since 0 may refer to an actual sequence number of the table (although in v1 all sequence number will be 0 so return 0 for v1 is better, but I'm not sure how easy it is to check table version here)

Reply:
Good catch. Make sense to set it to -1 since its initial value is 0 and increase monotonically as this line shows https://github.com/apache/iceberg/blob/9edb1fabc0a4ad4c3f10625c4a15918bf5790f8b/core/src/main/java/org/apache/iceberg/TableMetadata.java#L356. Will post a new commit.